### PR TITLE
Added ScoreCalculator class to detect change points give metric passed

### PR DESCRIPTION
**Pull request by maxslamdunk**

### CODE REVIEW

File: roerich/algorithms/calc_metrics.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    def __init__(self, metric=None, func=None, periods=1, window_size=100, step=1, n_runs=1):
+        super().__init__(periods=periods, window_size=window_size, step=step, n_runs=n_runs)
+        self.metric = metric
+        self.func = func
+
+    def reference_test_predict(self, X_ref, X_test):
+
+        if self.metric == "EnergyDist":
+            n = X_ref.shape[0]
+            E = 2*pairwise_distances(X_ref, X_test, metric='euclidean') - pairwise_distances(X_test, metric='euclidean') - pairwise_distances(X_ref, metric='euclidean')
```
Comment (hushchyn-mikhail) [inline roerich/algorithms/calc_metrics.py:28]: @maxslamdunk, hi!
Let's implements this function like it was done for fd and mmn in https://github.com/HSE-LAMBDA/roerich/tree/main/roerich/scores

File: roerich/algorithms/calc_metrics.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+class ScoreCalculator(ChangePointDetectionBase):
+
+    def __init__(self, metric=None, func=None, periods=1, window_size=100, step=1, n_runs=1):
+        super().__init__(periods=periods, window_size=window_size, step=step, n_runs=n_runs)
+        self.metric = metric
+        self.func = func
+
+    def reference_test_predict(self, X_ref, X_test):
+
+        if self.metric == "EnergyDist":
```
Comment (hushchyn-mikhail) [inline roerich/algorithms/calc_metrics.py:26]: @maxslamdunk, "EnergyDist" -> "energy"

File: roerich/algorithms/calc_metrics.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            n = X_ref.shape[0]
+            E = 2*pairwise_distances(X_ref, X_test, metric='euclidean') - pairwise_distances(X_test, metric='euclidean') - pairwise_distances(X_ref, metric='euclidean')
+            return np.sum(E) / n ** 2
+        elif self.metric == "FrechetDist":
```
Comment (hushchyn-mikhail) [inline roerich/algorithms/calc_metrics.py:30]: "FrechetDist" -> 'fd'

File: roerich/algorithms/calc_metrics.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+from scipy.signal import argrelmax
+
+from .cpdc import ChangePointDetectionBase
+from roerich.scores.fd import frechet_distance
+from roerich.scores.mmd import maximum_mean_discrepancy
+
+
+class ScoreCalculator(ChangePointDetectionBase):
+
+    def __init__(self, metric=None, func=None, periods=1, window_size=100, step=1, n_runs=1):
```
Comment (hushchyn-mikhail) [inline roerich/algorithms/calc_metrics.py:19]: Add docstring with description of all input hyperparameters, please. For an example, https://github.com/HSE-LAMBDA/roerich/blob/main/roerich/algorithms/cpdc_cv.py

File: roerich/algorithms/calc_metrics.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+from joblib import Parallel, delayed
+from scipy import interpolate
+from scipy.signal import argrelmax
+
+from .cpdc import ChangePointDetectionBase
+from roerich.scores.fd import frechet_distance
+from roerich.scores.mmd import maximum_mean_discrepancy
+
+
+class ScoreCalculator(ChangePointDetectionBase):
```
Comment (hushchyn-mikhail) [inline roerich/algorithms/calc_metrics.py:17]: ScoreCalculator -> SlidingWindows

File: roerich/algorithms/calc_metrics.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
+from .cpdc import ChangePointDetectionBase
```
Comment (hushchyn-mikhail) [inline roerich/algorithms/calc_metrics.py:1]: @maxslamdunk, move this file into https://github.com/HSE-LAMBDA/roerich/tree/main/roerich/change_point folder, please. And remane it as windows.py, for an example.

File: roerich/algorithms/calc_metrics.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+
+class SlidingWindows(ChangePointDetectionBase):
+
+    def __init__(self, metric=None, periods=1, window_size=100, step=1, n_runs=1):
+        super().__init__(periods=periods, window_size=window_size, step=step, n_runs=n_runs)
+        self.metric = metric
+
+        """
+        Change point detection algorithm based on binary classification.
```
Comment (hushchyn-mikhail) [inline roerich/algorithms/calc_metrics.py:14]: Not classification

File: roerich/algorithms/calc_metrics.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+        step: int, default=1
+            Algorithm estimates change point detection score for each <step> observation. step > 1 helps
+        to speed up the algorithm.
+
+        n_runs: int, default=1
+            Number of times, the binary classifier runs on each pair of test and reference
+        windows. Observations in the windows are divided randomly between train and validation sample every time.
+        n_runs > 1 helps to reduce noise in the change point detection score.
+        
+        metric: str/function, default=None
```
Comment (hushchyn-mikhail) [inline roerich/algorithms/calc_metrics.py:37]: metric must be on the first place

Comment (hushchyn-mikhail) [reply roerich/algorithms/calc_metrics.py:37]: And list all possible string values like metrics: {'energy', 'fd', 'mmd'}

File: roerich/algorithms/calc_metrics.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        if self.metric == "energy":
+            return energy_distance(X_ref, X_test)
+        elif self.metric == "fd":
+            return frechet_distance(X_ref, X_test)
+        elif self.metric == "mmd":
+            return maximum_mean_discrepancy(X_ref, X_test)
+        elif callable(self.metric):
+            return self.metric(X_ref, X_test)
+        else:
+            raise ValueError("metric should be one of: EnergyDist, FrechetDist, MaxMeanDisc; or a function should be "
```
Comment (hushchyn-mikhail) [inline roerich/algorithms/calc_metrics.py:67]: Change function names here, please

File: roerich/algorithms/calc_metrics.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        n_runs > 1 helps to reduce noise in the change point detection score.
+        
+        metric: str/function, default=None
+            Function that gives the measure of dissimilarity between data points in windows.
+        Metric should be one of: EnergyDist, FrechetDist, MaxMeanDisc; or a function should be passed.
+        Function must be in the following format:
+
+            Parameters:
+            -----------
+            X_ref: numpy.ndarray
```
Comment (hushchyn-mikhail) [inline roerich/algorithms/calc_metrics.py:44]: This is not for __init__. Remove it, please

File: roerich/algorithms/calc_metrics.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+            Number of consecutive observations of a time series in test and reference
+        windows. Recommendation: select the value so that there is only one change point within 2*window_size
+        observations of the signal.
+
+        step: int, default=1
+            Algorithm estimates change point detection score for each <step> observation. step > 1 helps
+        to speed up the algorithm.
+
+        n_runs: int, default=1
+            Number of times, the binary classifier runs on each pair of test and reference
```
Comment (hushchyn-mikhail) [inline roerich/algorithms/calc_metrics.py:33]: Not classififer

File: roerich/algorithms/calc_metrics.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+        windows. Observations in the windows are divided randomly between train and validation sample every time.
+        n_runs > 1 helps to reduce noise in the change point detection score.
+        
+        metric: str/function, default=None
+            Function that gives the measure of dissimilarity between data points in windows.
+        Metric should be one of: EnergyDist, FrechetDist, MaxMeanDisc; or a function should be passed.
+        Function must be in the following format:
```
Comment (hushchyn-mikhail) [inline roerich/algorithms/calc_metrics.py:40]: Just provide the function example. metric = roerich.scores.frechet_distance

File: roerich/algorithms/cpdc.py [hunk 13]
```diff
@@ -115,7 +115,7 @@ def reference_test_predict(self, X_ref, X_test):
         X_test: numpy.ndarray
             Matrix of test observations.
 
-        Retunrs:
+        Returns:
```
Comment (hushchyn-mikhail) [inline roerich/algorithms/cpdc.py:118]: No changes here, please.

File: roerich/scores/energy.py [hunk 14]
```diff
@@ -0,0 +1,14 @@
+import numpy as np
+from sklearn.metrics import pairwise_distances
+
+
+def energy_distance(x, y):
+    x = np.reshape(x, (len(x), -1))
+    y = np.reshape(y, (len(y), -1))
+    n = x.shape[0]
+    print(x.shape)
```
Comment (hushchyn-mikhail) [inline roerich/scores/energy.py:9]: @maxslamdunk, remove prints, please

File: roerich/change_point/windows.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+from roerich.scores.fd import frechet_distance
+from roerich.scores.mmd import maximum_mean_discrepancy
+from roerich.scores.energy import energy_distance
+import numpy as np
+from sklearn.utils import resample
+
+
+class SlidingWindows(ChangePointDetectionBase):
+
+    def __init__(self, metric='mmd', bootstrap=False, periods=1, window_size=100, step=1, n_runs=1):
```
Comment (hushchyn-mikhail) [inline roerich/change_point/windows.py:11]: Add bootstrap parameter description in the docstring, please
